Whole-array reductions always use init to start each reduction chain

mbauman opened the pull request:
This implements the new `init` behaviors that were decided in #53945 — that `init` is used one or more times to start each reduction chain.  Unlike the dimensional counterpart (#55318), this can be done without a big refactor, which makes it much easier to evaluate the what the core behaviors should be.  And, importantly, it makes it much easier for bigger refactorings — like #52397 — to be implemented subsequently.  Perhaps most importantly, it's what we need to do in order to officially document that this is how `init` works.  The current docs give us leeway to do this, and then once it's implemented, we can document the better semantics (and finally merge #53945).

In implementing this, I relalized there's still an outstanding fundamental question on how reductions should work.  #53945 clarifies how `init` itself is used, but there's still an outstanding question about what needs to happen *next*.  

The crux of it is how reductions should start.  I talk in #55318 about the 2*3 different ways in which a reduction may start... and here I codified it all under a single function:

```Julia
"""
    _mapreduce_start(f, op, A, init, [a1], [a2])

Perform the first step in a mapped reduction over `A` with 0, 1, or more elements.
The two element method may be called multiple times within a single reduction at
the start of each new chain of `op` calls.
"""
_mapreduce_start(f, op, A, ::_InitialValue) = mapreduce_empty(f, op, eltype(A))
_mapreduce_start(f, op, A, ::_InitialValue, a1) = mapreduce_first(f, op, a1)
_mapreduce_start(f, op, A, ::_InitialValue, a1, a2) = op(f(a1), f(a2))
_mapreduce_start(f, op, A, init) = init
_mapreduce_start(f, op, A, init, a1) = op(init, f(a1))
_mapreduce_start(f, op, A, init, a1, a2) = op(op(init, f(a1)), f(a2))
```

This brings all this chicanery into one consolidated place — and makes it clear why you need 6 cases when these are the definitions.

But there's (at least) one very interesting test that breaks, and it makes for a perfect example (just ignore the fact that it breaks the "neutral element" init rule; that's actually *not* the interesting part here!)

```
julia> @inferred count(false:true, dims = (:), init = 0x0004) === 0x0005
ERROR: return type UInt16 does not match inferred return type Union{UInt16, UInt64}
```

On master, Julia just happily returns a _properly inferred_ 0x005.  Why did this change?  Well, now we can use an `init` in the pairwise implementation, so this now uses the pairwise implementation (hooray!). But now that adds an additional possible branch here.  This could be computed (and master does so) as the foldl `(0x0004 + false) + true`, where you only ever add `UInt16` and `Bool`.  Or (with more elements) you could do `((0x0004 + false) + ...) + ((0x0004 + true) + ...)`, and that final summation is between two `UInt16`.  And I lied, this isn't computed with `+`, it's `add_sum`, which very intentionally promotes that case (but not the other) to `UInt`.  So there's the instability.

But is this actually how we want reductions to start?  What if we **always called `mapreduce_first`**?  I had been thinking it was the analog of `empty` — in that it only gets called when the entire reduction is just 1 element — and that's also how @stevengj had implemented the streaming iterable pairwise algorithm as well.  But it's actually documented that it "may also be used to initialise the recursion."  In other words, maybe the above `_mapreduce_start` definition should be — more simply:

```Julia
_mapreduce_start(f, op, A, ::_InitialValue) = mapreduce_empty(f, op, _empty_eltype(A))
_mapreduce_start(f, op, A, ::_InitialValue, a1) = mapreduce_first(f, op, a1)
_mapreduce_start(f, op, A, ::_InitialValue, a1, a2) = op(mapreduce_first(f, op, a1), f(a2))
_mapreduce_start(f, op, A, init) = init
_mapreduce_start(f, op, A, init, a1) = op(init, mapreduce_first(f, op, a1))
_mapreduce_start(f, op, A, init, a1, a2) = op(op(init, mapreduce_first(f, op, a1)), f(a2))
```

And then you squint at it, and you realize the two-element case is no longer special — it's just the natural "next step" of the reduction after the one-element starting point in both of the init-or-not cases!  This is great, as only doing something special to _one_ element before launching into a for loop or recursion is drastically (and perhaps surprisingly so) simpler than needing to do something special to _two_ elements.  It removes a case from every single implementation, and every single implementation only needs to peel one element out of a loop instead of two.  It also fixes one bit of breakage that I initially added here — that `mapreduce_impl` requires at least two elements with the previous definitions, but it only requires one element on master.

Now the test above _is still broken_, but it's broken in a new and more interesting way:

```Julia
julia> @inferred count(false:true, dims = (:), init = 0x0004)
ERROR: return type Int64 does not match inferred return type Union{Int64, UInt16}
```

There are really only a handful of meaningful things that `mapreduce_first` does, but one of them is to promote `Bool`s in summations to `Int` — and `count` is just a `sum` of bools.  So now the instability is *just* stemming from our choice of `init` itself.

So that's now the question: is that a better behavior?  Maybe?  I've gone through a handful of tests and stdlibs locally, but I'm curious what CI (and eventually PkgEval) says. 

This fixes #47216.

mbauman committed:
consistently use `init` in whole-array mapreduce

mbauman committed:
TEMPORARILY USE STDLIB BRANCHES

mbauman committed:
better error message for no eltype

mbauman committed:
always use reduce_first?

mbauman committed:
fix tests

mbauman committed:
`count`'s init is initialization, not type assertion

adienes commented:
part of me feels like this is a problem with `count` and not `mapreduce`

> On master, Julia just happily returns a properly inferred 0x005.

is this "properly"?  I could totally see an argument that `count` should always return `Int` regardless of init type.

mbauman commented:
It is about mapreduce at the end of the day; `count` is just the (only!) canary that's squawking.  The core behavior change is in both `sum([false,true], init=0x00)` and `mapreduce(identity, +, [false,true], init=0x00)`.

There are two orthogonal features of the reduction machinery at play.  They're also fundamentally _where_ the problem originates in the first place.

* We want `sum` to avoid integer overflow.  So it uses a special `Base.add_sum` function as its reduction `op`.  This makes small signed integers (but not including `Bool`, although `::Bool + ::Bool` itself promotes to `Int`) promote to `Int`.  Perhaps `Bool` should be included here.  That would solve the problem, but that doesn't help us answer what `mapreduce_first` should do.
* We want reductions of one element to be type stable when `init` is not provided — but they cannot call `op` directly.  So `mapreduce_first` gives `op`s the *op*portunity to specify what should happen.  In the case of `add_sum`, it promotes `Bool` *and* other small signed integers to `Int`.  But what about other cases — when `init` is provided and/or there's more than one element?  If `mapreduce_first` was unconditionally called, then that would _also_ solve the problem.  In fact, we'd no longer need to define methods on the `add_sum` function itself; this would be sufficient to get the promoting behaviors we wanted.  If `mapreduce_first` solves the _output_ type stability problem, then that means (definitionally) that it also solves the _inner_ type stability of the reduction variable itself.

I've always thought of `mapreduce_first` as being solely about _output_ type stability with one element.  That indeed was its original raison d'être (https://github.com/JuliaLang/julia/pull/25051). But if it does good work there, why _shouldn't_ we also use it at the beginning of each reduction chain — particularly since the possibility of its use to initialize a reduction chain has _always been_ a part of its docs!

mbauman commented:
It's still taking me some time to actually figure out what this is doing, but I think I'm getting convinced myself, too.  This example works pretty well at expressing all the possibilities in a single concise table.

On Julia v1.12/current master, here's the representative universe of what `sum` does:

```julia
using Base: reduce_empty as r0, reduce_first as r1, add_sum as +ₛ
const t, f = true, false
# #= high-level reduc =# === #= internal API =#   === #= expand r/+ₛ +# === #= value =#
sum(Bool[])              === r0(+ₛ, Bool)         === 0
sum([f])                 === r1(+ₛ, f)            === 0
sum([f,t])               ===  f +ₛ t              ===  f + t            === 1
sum([f,t,f,t])           === (f +ₛ t) +ₛ (f +ₛ t) === (f + t) + (f + t) === 2
sum(Bool[],   init=0x00) ===  0x00
sum([f],      init=0x00) ===  0x00 +ₛ f           === 0x00 + f          === 0x01
sum([f,t],    init=0x00) === (0x00 +ₛ f) +ₛ t     === (0x00 + f) + t    === 0x01
sum([f,t,f,t],init=0x00) === (((0x00 +ₛ f) +ₛ t) +ₛ f) +ₛ t === (((0x00 + f) + t) + f) + t === 0x02
```

If we jam in a pairwise implementation with init like I tried at first, we get that _extra_ promotion from the step that "combines" the two reduction chains.  This is the _only place_ where the `add_sum` function call itself does something different from `+` in this entire post:

```julia
sum([f,t,f,t],init=0x00) === ((0x00 +ₛ f) +ₛ t) +ₛ ((0x00 +ₛ f) +ₛ t) === UInt(0x00 + f + t) + UInt(0x00 + f + t) === UInt(0x01) + UInt(0x01) === UInt(2)
```

The current state of this PR calls `reduce_first` in every single case except the empty ones, so the whole scheme becomes:

```julia
using Base: reduce_empty as r0, reduce_first as r1, add_sum as +ₛ
const t, f = true, false
# #= high-level reduc =# === #= internal implementation =#         === #= expanding r0/r1/+ₛ =#
sum(Bool[])              === r0(+ₛ, Bool)                          === 0
sum([f])                 === r1(+ₛ, f)                             === 0
sum([f,t])               === r1(+ₛ, f) +ₛ t                        ===  0 + t            === 1
sum([f,t,f,t])           === (r1(+ₛ, f) +ₛ t) +ₛ (r1(+ₛ, f) +ₛ t)  === (0 + t) + (0 + t) === 2
sum(Bool[],   init=0x00) ===   0x00
sum([f],      init=0x00) ===   0x00 +ₛ r1(+ₛ, f)                   ===  0x00 + 0      === 1
sum([f,t],    init=0x00) ===  (0x00 +ₛ r1(+ₛ, f)) +ₛ t             === (0x00 + 0) + t === 1
sum([f,t,f,t],init=0x00) === ((0x00 +ₛ r1(+ₛ, f)) +ₛ t) +ₛ ((0x00 +ₛ r1(+ₛ, f)) +ₛ t) === (0x00 + 0 + t) + (0x00 + 0 + t) === 2
```

This probably makes it _easier_ to ensure subsequent type stability, because the type of `r1` is _always_ a part of the reduction.

---------------

It's probably worth noting that these effects are quite limited — there simply aren't many `reduce_first` definitions in the first place.  Base only defines these 9 methods, and there's only a [handful of packages in the ecosystem](https://juliahub.com/ui/Search?type=symbols&q=reduce_first&u=define) that define similar sum/prod methods.

```julia
reduce_first(op, x) = x
reduce_first(::typeof(+), x::Bool) = Int(x)
reduce_first(::typeof(*), x::AbstractChar) = string(x)
reduce_first(::typeof(add_sum), x) = reduce_first(+, x)
reduce_first(::typeof(add_sum), x::BitSignedSmall)   = Int(x)
reduce_first(::typeof(add_sum), x::BitUnsignedSmall) = UInt(x)
reduce_first(::typeof(mul_prod), x) = reduce_first(*, x)
reduce_first(::typeof(mul_prod), x::BitSignedSmall)   = Int(x)
reduce_first(::typeof(mul_prod), x::BitUnsignedSmall) = UInt(x)
```

mbauman commented:
Let's see what Nanosoldier thinks.  I can still make this less breaking — most notably by relaxing the `mapreduce_impl`'s requirement to have two elements back to one again with that `reduce_first` simplification — but that's all internalsy stuff anyhow.  I'm more curious about the behavior changes.

@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/58241#issuecomment-2836807764) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**1 packages crashed only on the current version.**

 - A segmentation fault happened: 1 packages

6 packages crashed on the previous version too.

## ✖ Packages that failed

**82 packages failed only on the current version.**

 - Package fails to precompile: 16 packages
 - Package has test failures: 9 packages
 - Package tests unexpectedly errored: 42 packages
 - Tests became inactive: 1 packages
 - Test duration exceeded the time limit: 12 packages
 - Test log exceeded the size limit: 2 packages

1487 packages failed on the previous version too.

## ✔ Packages that passed tests

**18 packages passed tests only on the current version.**

 - Other: 18 packages

5183 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**13 packages successfully loaded only on the current version.**

 - Other: 13 packages

2810 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

917 packages were skipped on the previous version too.



</details>

mbauman commented:
Well there are still a number of meaningful errors here, from most interesting to least:

* [OnlineStats](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/OnlineStats.primary.log) calls `reduce(merge!, ks, init = KahanSum(T))`, where a single mutable `init` is updated and then re-used multiple times.
* [AcceleratedKernels](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/AcceleratedKernels.primary.log) calls `AK.reduce(+, v; init=Int32(10))`. `AK.reduce` is indeed separate from Base, but it falls back to `Base.reduce` for CPU backends. [AcceleratedKernels.jl/test/runtests.jl](https://github.com/JuliaGPU/AcceleratedKernels.jl/blob/0b99fbf767ff8c6f7bbd862082c34b178f1189ce/test/runtests.jl#L721 )
* [Gridap](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/GridapGmsh.primary.log) (and deps) have this strange promotion thing going sideways in `count`. In short, Gridap defines a type `VectorValue` that's a subtype of `Number` but is itself iterable and may have more than one element.  They run `count` over these `VectorValue`s, expecting to _iterate_ over the _many values_ in the Number.  On master, this _misses_ the `mapreduce(f, op, ::Number)` specialization because `count` calls `sum` with an `init`.  This PR brings `init` support to that `::Number` specialization, which then doesn't iterate multiple values, and then leads to this not working as Gridap expected. 
* [DistributedArrays](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/DistributedArrays.primary.log) overloads the internal `Base._mapreduce` and `Base.mapreducedim!` in ways currently incompatible with this PR and loses their specializations
* [InMemoryDatasets](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/InMemoryDatasets.primary.log) directly calls the internal `mapreduce_impl` in `src/stat/non_hp_stat.jl:165`, leading to about a half dozen failures.
* The vast majority of failures are because Nanosoldier didn't pick up the Statistics branch here, and Statistics directly calls the internal `mapreduce_impl`
* A `preprocess`ed png image isn't bitwise identical in [TensorBoardLogger](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/TensorBoardLogger.primary.log)? 
* Approximate tolerances exceeded in [CopEnt](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/CopEnt.primary.log) and [DirectTrajOpt](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/DirectTrajOpt.primary.log) and [AnyMOD](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/AnyMOD.primary.log)
* [StateSpaceDynamics](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/StateSpaceDynamics.primary.log) failed to converge
* [TravelingSalesmanExact](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/TravelingSalesmanExact.primary.log) ended up with an empty solutions vector somehow
* [EarlyStopping](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8d6b63c_vs_8a8e3d1/EarlyStopping.primary.log) looks unrelated

I think I can fiddle around with some of the internal method signatures to make this a little more kind to the folks playing with fire.

adienes commented:
while clever, `reduce(merge!, ks, init = KahanSum(T))` does not feel legal

mbauman commented:
Yeah it's definitely not. They also test `foldl` (which is ok!) and I also get the answer they expect if I make a slew of copies with `reduce((x,y)->merge!(copy(x), y), ks, init = KahanSum(T))` or `reduce(merge!, copy.(ks))`.  It makes you want for an `init=()->Kahan(T)` sort of thing though, doesn't it?

It feels like you could almost make this reliable with `reduce_first` — and looking back now, this is what's discussed in https://github.com/JuliaLang/julia/issues/47696:

```julia
julia> foldl(merge!, ks, init = KahanSum(T)) |> sum
250439.49035165994

julia> ks_reduce = reduce((x,y)->merge!(copy(x), y), ks, init = KahanSum(T)) |> sum
250439.49035165994

julia> Base.reduce_first(::typeof(merge!), k::KahanSum) = copy(k) # prevent mutating the input array without init

julia> reduce(merge!, ks) |> sum
250439.49035165994
```

The trouble is, though, that it doesn't solve this footgun:

```julia
julia> reduce(merge!, ks, init=KahanSum(T)) |> sum
8.360444161382082e18
```

I don't really see a simple solution; I think we'd have to promise _so much_ about the internals in order to have a sensible way to rely on mutation... and even then I think it'd be quite fragile.

mbauman committed:
simplify reduction start to 2*2 cases: (empty/not) * (init/not)

mbauman committed:
less disruptive internal method argument orderings

mbauman committed:
Revert "TEMPORARILY USE STDLIB BRANCHES"

This reverts commit c07a5f7cbb2fd31f28c7d594f914825e0b13f547. The Statistics branch is no longer needed; the Sparse branch is still beneficial but not required.

mbauman committed:
fixup! less disruptive internal method argument orderings

mbauman commented:
@nanosoldier `runtests(["Statistics", "NonConvexPenalizedRegression", "EarlyStopping", "StandardizedMatrices", "BarkerMCMC", "BoltzmannMachines", "DistributedArrays", "Normalization", "NaNStatistics", "FeatureSelection", "TimeSeries", "OnlineStatsBase", "CoupledFields", "InMemoryDatasets", "StatsBase", "LongDatasetSort", "CartesianJoin", "TimeSeriesEcon", "DLMReader", "CopEnt", "ASCertain", "QuantumPropagators", "GraphicalLasso", "MLJScikitLearnInterface", "OnlineStats", "SpatialDependence", "TensorBoardLogger", "TransitionalMCMC", "LaTeXCompilers", "FastDMTransform", "VectorizedStatistics", "SDeMo", "Tonari", "MDEStudy", "AcceleratedKernels", "PopGenCore", "MFCC", "Pathogen", "GlobalSensitivityAnalysis", "Lasso", "TravelingSalesmanExact", "DMRGenie", "DirectTrajOpt", "SDDP", "HurdleDMR", "GtkObservables", "FlashWeave", "GenomicOffsets", "LiftAndLearn", "TaylorIntegration", "IRKGaussLegendre", "StateSpaceDynamics", "StochasticSeriesExpansion", "AnyMOD", "StatisticalGraphics", "EquilibratedFlux", "DynamicHMC", "RobustModels", "BulkLMM", "PopGen", "GridapDistributed", "GenX", "TinyGibbs", "BoltzmannMachinesPlots", "GridapPETSc", "PosteriorStats", "MCMCChains", "GridapP4est", "NestedSamplers", "GridapSolvers", "OpticalFibers", "DynamicHMCExamples", "GlobalSensitivity", "GridapGmsh", "ReactionNetworkImporters", "JumpProblemLibrary", "DataDrivenSparse", "Chron", "ModiaLang", "Spectra", "GridapMakie", "MaterialPointVisualizer", "BloqadeNoisy"])`

mbauman committed:
review uses of init

there are still tests that use "illegal" init values but happen to be ok due to the number of elements or somesuch; I am intentionally leaving them as they may be helpful to flag if and when further refactorings change these fraught behaviors.

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/58241#issuecomment-2846062823) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/0d3c9b0_vs_8a8e3d1/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**1 packages crashed only on the current version.**

 - A segmentation fault happened: 1 packages


## ✖ Packages that failed

**15 packages failed only on the current version.**

 - Package fails to precompile: 7 packages
 - Package has test failures: 3 packages
 - Package tests unexpectedly errored: 2 packages
 - Test duration exceeded the time limit: 3 packages

8 packages failed on the previous version too.

## ✔ Packages that passed tests

**2 packages passed tests only on the current version.**

 - Other: 2 packages

53 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**1 packages successfully loaded only on the current version.**

 - Other: 1 packages

3 packages successfully loaded on the previous version too.



</details>

mbauman commented:
Great, that's much better.  We still have Gridap's `Number` of length > 1, OnlineStats + AcceleratedKernel's non-neutral inits in their tests, as I'd expect.  But we also still have the few less-obvious and harder to identify failures to converge in AnyMOD ([looking for a specific objective value](https://github.com/leonardgoeke/AnyMOD.jl/blob/dede89b9e9875d4a4c0abb3cd0ef401cc724c46a/test/runtests.jl#L31)), traveling salesman (getting missing solutions in [att48.tsp](https://github.com/ericphanson/TravelingSalesmanExact.jl/blob/5804bb37c8d781da4d3dcfc8b9d7446ac87adab6/test/runtests.jl#L159)) and StateSpaceDynamics getting non-finite values in some matrix somehow ([here](https://github.com/depasquale-lab/StateSpaceDynamics.jl/blob/67707fd1add7813f745e79111424daac5debc0f0/test/LinearDynamicalSystems/PoissonLDS.jl#L243)).
